File: examples/asciiart/asciiart.js
/* AsciiArt - Outputs an ascii dungeon layout */
const hre = require("hardhat");
const ethers = hre.ethers;

const id = 8032;    // Change this to your dungeon ID

const dungeonsAddress = "0x86f7692569914B5060Ef39aAb99e62eC96A6Ed45"

// Formatting for console output 
const cyanStart = '\x1b[36m\x1b[1m'
const cyanEnd = '\x1b[22m\x1b[0m'
const magentaStart = '\x1b[35m\x1b[1m'
const magentaEnd = '\x1b[22m\x1b[0m'

async function main() {

    const Dungeons = await ethers.getContractFactory('Dungeons');
    const dungeons = await Dungeons.attach(dungeonsAddress);

    const name = await dungeons.getName(id);
    const layout = await dungeons.getLayout(id);
    const size = await dungeons.getSize(id);
    const entities = await dungeons.getEntities(id);

    // Parse Layout into binary 
    layoutInt = BigInt(layout)              // Process uint256 -> javascript readable int
    let bits = layoutInt.toString(2);       // Convert BigInt to binary)
    let length = Math.ceil(bits.length / 256)*256;  // Bits are stored in chunks of 256. Figure out how many chunks we have.
    bits = bits.padStart(length, '0');              // Add padding at the start to be a full '256' or '512'

    // Store dungeon in 2D array
    dungeon = []; 

    let counter = 0;
    for(let y = 0; y < size; y++) {
        let row = []
        for(let x = 0; x < size; x++) {
            const bit = bits[counter] == 1 ? ' ' : 'X';
            row.push(bit)
            counter++;
        }
        dungeon.push(row)
    }  

	// Parse entities into array 
	let entityList = []

    // Make sure we have entities to parse
	if(entities[0].length > 0) {	
		for(let i = 0; i < entities[0].length; i++) {
			let entity = {
				x: entities[0][i],
				y: entities[1][i],
				entityType: entities[2][i]
			}

			entityList.push(entity)

            // Update dungeon with our entity
            if(entity.entityType == 1) {
                // Place a door
                dungeon[entity.y][entity.x] = `${magentaStart}p${magentaEnd}`
            } else if(entity.entityType == 0) {
                // Place a point of interest
                dungeon[entity.y][entity.x] = `${cyanStart}D${cyanEnd}`
            } 
		}
	}
    
    console.log(`#${id} ${name}`);
    console.log(toString(dungeon));
}

function toString(dungeon) {
  // Returns a string representing the dungeon
  let rowString = ""

  for(let y = 0; y < dungeon.length; y++) {
      for(let x = 0; x < dungeon.length; x++) {
          const tile = dungeon[y][x]
          rowString += `${tile} `
      }
      rowString += '\n'
  }
  return(rowString)
}



main()
  .then(() => process.exit(0))
  .catch((error) => {
    console.error(error);
    process.exit(1);
  });

File: examples/asciiart/asciiart.py
# AsciiArt - Outputs an ascii dungeon layout
# Python equivalent to asciiart.js

from etherscan import Etherscan
from web3 import Web3
from web3.auto import w3
from web3.auto.infura import w3
import os
import math

contract_address = "0x86f7692569914B5060Ef39aAb99e62eC96A6Ed45"
w3 = Web3(Web3.HTTPProvider('https://mainnet.infura.io/v3/9f420f93b67041a5866c27bcb1c30659')) # Infura endpoint
INFURA_API_KEY = "9f420f93b67041a5866c27bcb1c30659" # Infura api key
api_key = "ITB9PRB1FHNQWHQI8GXU5UU463D8G8KX92"# Etherscan api key
id = 8032 #Dungeon ID

eth = Etherscan(api_key)
abi = eth.get_contract_abi(contract_address)
Contract = w3.eth.contract(address=contract_address, abi=str(abi))

# Call contract functions
layout = Contract.functions.getLayout(id).call()
entities = Contract.functions.getEntities(id).call()
size = Contract.functions.getSize(id).call()
name  = Contract.functions.getName(id).call()

# Convert layout into binary
hex_layout = Web3.toHex(layout)
int_value = int(hex_layout, 16) # Convert to int
bin_value = bin(int_value)[2:] # Convert to binary
length = math.ceil(len(bin_value)/256) * 256
bins = bin_value.zfill(length) # Add padding at the start to be a full '256' or '512'

y = range(size)
x = range(size)

# Store dungeon in 2D array
dungeon = []

counter = 0
for i in y:
    row = []
    for i in x:
        bit = int(bins[counter])
        if bit == 1:
            row.append("  ")
        else:
            row.append('X ')
        counter +=1
    dungeon.append(row)

# Parse entities into array
entitylist = []
ent = len(entities[0])
e = range(ent)

if ent > 0:
    for i in e:
        entity = [entities[1][i], entities[0][i], entities[2][i]]
        entitylist.append(entity)

    for i in entitylist:
        # Place a point
        if i[2] == 0:
            dungeon[i[0]][i[1]] = "P "
        # Place a door
        elif i[2] == 1:
            dungeon[i[0]][i[1]] = "D "

def to_string(dungeon):
    # Returns a string representing the dungeon
    rowString = ""
    for i in y:
        for t in x:
            tile = dungeon[i][t]
            rowString += tile
        rowString += '\n'
    return(rowString)

print(f"#{id} {name}")
print(to_string(dungeon))
